working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+connected components).  For the example above, C and D are condensed into a
+single strongly connected component, so the graph now looks like this:
+
+[guaranteed to be acyclic]: https://en.wikipedia.org/wiki/Strongly_connected_component#Definitions
+
+&emsp;(B) &lArr; (A) &rArr; (C &hArr; D)
+
+Then, for each stage of inference, we find all the strongly connected components
+with no graph edges pointing to them, and remove them from the graph.  We repeat
+the process for successive stages, until no strongly connected components remain
+in the graph.  The set of arguments we will recursively infer during each stage
+is the union of the arguments in all the strongly connected components removed
+from the graph during that stage.
+
+In this example, that means that there will be two stages.  The first stage will
+consist of argument A, and the second stage will consist of arguments B, C, and
+D.
+
+(Note that the invariant mentioned earlier, that non-function literals are
+always placed in the first stage, is guaranteed by the fact that dependency
+analysis only draws an edge from _A_ to _B_ if _A_ is a function literal.)
+
+### What constitutes a function literal?
+
+The determination of whether an argument is a function literal is made
+syntactically.  An argument is a function literal if it is produced by one of
+the following grammar productions (as described in the language specification):
+
+* _\<functionExpression\>_
+
+* _\<namedArgument\>_, where the _\<expression\>_ part is a function literal.
+
+* `(` _\<expression\>_ `)`, where the _\<expression\>_ part is a function
+  literal.
+
+### Why is it safe to alter the visit order?
+
+Note that in most cases, flow analysis depends critically on the fact that type
+inference visits subexpressions in the same order they are evaluated.  For
+example, in the following code, the first argument to `f` requires a null check,
+because `int?` does not support `operator +`.
+
+```dart
+void f(int i, int j) { ... }
+void g(int? i) {
+  f(i! + 1, i = 0);
+}
+```
+
+If type inference instead visited the `i = 0` argument first, the `!` would be
+considered unnecessary (because `i = 0` promotes the type of `i` to non-nullable
+`int`), so we would be permitted to write `f(i + 1, i = 0)`.  However, this
+would fail at runtime, because arguments are evaluated in source order, so `i +
+1` is evaluated before `i = 0`.
+
+In general, we risk these sorts of breakages any time we change the order in
+which we perform type inference on invocation arguments.  However, there's an
+exception for arguments that are function literals: they may be safely
+re-ordered within the invocation, because the act of _evaluating_ a function
+literal doesn't actually have any runtime effect.  The runtime effect doesn't
+happen until the function literal is _called_, and the earliest this can
+possibly happen until after all the invocation arguments have been evaluated.
```

Comment:
Yes, that's true.  Flow analysis understands that, upon entry to the top of the `while` loop, any write captures that occur inside the loop may already have occurred (because the loop may execute more than once), so it considers any write captures inside the loop itself to take effect at the top of the loop.